Signal/src/network/GiphyDownloader.swift

```diff
@@ -617,19 +692,31 @@ extension URLSessionTask {
         }
     }
 
+    // Return the first asset request for which we either:
+    //
+    // * Need to download the content length.
+    // * Need to download at least one of its segments.
     private func popNextAssetRequest() -> GiphyAssetRequest? {
         AssertIsOnMainThread()
 
+        let kMaxAssetRequestCount: UInt = 6
+        let kMaxAssetRequestsPerAssetCount: UInt = 2
+
         // Prefer the first "high" priority request;
         // fall back to the first "low" priority request.
-        var activeAssetRequestsCount = 0
+        var activeAssetRequestsCount: UInt = 0
         for priority in [GiphyRequestPriority.high, GiphyRequestPriority.low] {
             for assetRequest in assetRequestQueue where assetRequest.priority == priority {
                 switch assetRequest.state {
                 case .waiting:
-                    break
+                    // This asset request needs its content length.
+                    return assetRequest
                 case .requestingSize:
                     activeAssetRequestsCount += 1
+                    // Ensure that only N requests are active at a time.
+                    guard activeAssetRequestsCount < kMaxAssetRequestCount else {
```

Comment:
Naming suggests this should be `<=`

Reply:
It seems like this guard belongs in the `case .waiting` instead of here.

I think we want to do what we can to ensure something of a FIFO priority with our GiphyAssets. And requesting content size is a way to insert network requests for Asset which we haven't "started" yet.

Reply:
> Naming suggests this should be <=

We're try to find an asset for which we can add a request.  If we're at MAX, we don't want to add another request and can exit early.

Reply:
> It seems like this guard belongs in the case .waiting instead of here.

We can abandon our search for a request to start as soon as we hit the limit.  We want this method to be cheap since we call it often.

> I think we want to do what we can to ensure something of a FIFO priority with our GiphyAssets. 

We do honor FIFO, within the limitation that we have an upper bound on the number of requests per asset.  Are you saying we shouldn't do that?  

> And requesting content size is a way to insert network requests for Asset which we haven't "started" yet.

Right under the current arrangement we can have requests for multiple assets at a time.  For example, when the picker opens, it'll immediately do HEAD on the first six assets.  When the first HEAD completes, we'll start a segment download.  etc.

If you think we need a different arrangement let's video chat.